applications/IgaApplication/custom_elements/embedded_isogeometric_beam_element.cpp

```diff
@@ -0,0 +1,5629 @@
+    }
+
+
+    void EmbeddedIsogeometricBeamElement::GetElementOrientation(const Matrix& r_DN_De, const ConfigurationType& rConfiguration , Vector3d& B1, Vector3d& B2, Vector3d& B3, float& A, float& B)
+    {
+        const GeometryType& r_geometry = GetGeometry();
+        const SizeType number_of_nodes = r_geometry.size();
+
+        array_1d<double, 3> tangents;
+        GetGeometry().Calculate(LOCAL_TANGENT, tangents); //this returns the tangents of the embedded beam in the parameter space of the shell
+
+        const SizeType dimension = GetGeometry().WorkingSpaceDimension();
+        array_1d<double, 3> actual_base_vector = ZeroVector(dimension);
+
+        Vector current_displacement = ZeroVector(dimension * number_of_nodes);
+        if (rConfiguration == ConfigurationType::Current) GetValuesVector(current_displacement);
+
+        //basis vectors a1 and a2
+        Vector A1 = ZeroVector(dimension);
+        Vector A2 = ZeroVector(dimension);
+
+        for (SizeType i = 0;i < number_of_nodes;++i) {
+            A1[0] += (GetGeometry().GetPoint(i).X0() + current_displacement[i * dimension]) * r_DN_De(i, 0);
+            A1[1] += (GetGeometry().GetPoint(i).Y0() + current_displacement[(i * dimension) + 1]) * r_DN_De(i, 0);
+            A1[2] += (GetGeometry().GetPoint(i).Z0() + current_displacement[(i * dimension) + 2]) * r_DN_De(i, 0);
+
+            A2[0] += (GetGeometry().GetPoint(i).X0() + current_displacement[i * dimension]) * r_DN_De(i, 1);
+            A2[1] += (GetGeometry().GetPoint(i).Y0() + current_displacement[(i * dimension) + 1]) * r_DN_De(i, 1);
+            A2[2] += (GetGeometry().GetPoint(i).Z0() + current_displacement[(i * dimension) + 2]) * r_DN_De(i, 1);
+        }
+
+        B1 = A1 * tangents[0] + A2 * tangents[1]; //calculates the tangential vector B1 or b1 according to eq.(6)
+        cross_prod(B2, A1, A2); //calculates the normal vector B2 or b2 according to eq. (7)
+        cross_prod(B3, B1, B2); //calculates the binormalvector B3 or b3 according to eq. (8)
+
+
+        A = norm_2(A1);  //length of the base vector
+        float tmp = inner_prod(A2, A2) - pow(inner_prod(A1, A2), 2) / pow(A, 2);
+        if (fabs(tmp) > Tol)
+        {
+            B = sqrt(tmp);
+        }
+        else
+            B = 0;
+
+    }
+
+    void EmbeddedIsogeometricBeamElement::parametric_mapping(Vector& _knot_vector, double  u, double & u_mid  )
+    {
+        for (size_t i = 0; i < _knot_vector.size() - 1; ++i) {
+            if (u >= _knot_vector(i) && u < _knot_vector(i + 1)) {
+                u_mid = 0.5 * (_knot_vector(i) + _knot_vector(i + 1));
+                return;
+            }
+        }
+        u_mid = -1.0;
+    }
+
+    void EmbeddedIsogeometricBeamElement::comp_T_var(Vector& _t_var, Vector& _deriv, Vector3d& _r1)
+    {
+        //const unsigned int N_Dof = 4;//this->GetGeometry().PointsNumber()* (this->GetGeometry().WorkingSpaceDimension() + 1);
+        _t_var.resize(3 * N_Dof);
+        _t_var.clear();
```

Comment:
add some docum because is not possible to follow what you are actually doing here